### DISCUSSION TITLE
SD to I2S MP3 file metadata causing issues.

### DISCUSSION THREAD

**Discussion by sugnaigiul** [pschatzmann/arduino-audio-tools#1756] [Q&A] [answered]
Hello !

I have recently been tinkering with the basics of this very well made library and have encountered a strange issue.

First of all my setup is an ESP32S3 from Seeed with the "Sense" shield (this is where my SD card is) linked to a pcm5102 via I2S.
![setup-audio](https://github.com/user-attachments/assets/b0ec0a12-ca39-498b-97d3-83f79bde07f6)

The main problem I was having is that some of the mp3 files that were stored on my SD card would play without issue and some other wouldn't and would reset the board any time they where played. Just to be clear, i was getting the mp3 files from some youtube to mp3 websites.
Here is what would usually happen in the serial when one of the "wrong" file would be played : 

[D] StreamCopy.h : 100 - copy 1024 bytes 
[D] StreamCopy.h : 243 - available: 9013609
[D] StreamCopy.h : 390 - write: 1024
[D] AudioEncoded.h : 187 - EncodedAudioOutput::write: 1024
[D] CodecMP3Helix.h : 112 - virtual size_t audio_tools::MP3DecoderHelix::write(const uint8_t*, size_t): 1024
[D] AudioEncoded.h : 199 - EncodedAudioOutput::write: 1024 -> 1024
[D] StreamCopy.h : 396 - write: 1024 -> 1024
[I] StreamCopy.h : 163 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[D] StreamCopy.h : 100 - copy 1024 bytes 
[D] StreamCopy.h : 243 - available: 9012585
[D] StreamCopy.h : 390 - write: 1024
[D] AudioEncoded.h : 187 - EncodedAudioOutput::write: 1024
[D] CodecMP3Helix.h : 112 - virtual size_t audio_tools::MP3DecoderHelix::write(const uint8_t*, size_t): 1024
----RESET SEEMED TO HAPPEN HERE----
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x8 (TG1WDT_SYS_RST),boot:0x29 (SPI_FAST_FLASH_BOOT)
Saved PC:0x4201a707
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3818,len:0x109c
...

At first i thought that the issue was coming from the size of the files but i quickly (after a few days) realized that it rather originated from the content of the meta data of the mp3 files.

When i downloaded mp3s from the web some of them would have extra info such as name of artist, album etc while some wouldn't.
The one that were working were only the one without that data :
![paintMp3](https://github.com/user-attachments/assets/b558b6fa-99ae-4865-9fa2-a41f017f5c56)

I then found that i could make any file that previously did not work function by removing its metadata, although not through the windows files property menu (that would make the extra lines disapear in the directory but the file would still not function) but rather by using some random website (i am not sure i am allowed to link it here but its the first one that pops up when you search "remove metadata form mp3").

I thought that this was all great but i still wanted to know if someone could give me some extra infromation so that i can be sure i understand the issue.

Thank you !

**Comment by pschatzmann** [comment, marked as answer]
MP3 decoders are usually resilient enough that they just ignore data they can not process and I have never had any issues with this so far.

Can you share one of the files that create problems ?
You can maybe also try another mp3 decoder...

If you want to contribute, we can try to extend the existing metadata parser to extract the audio, so that we can feed pure audio to the codec.

**Reply by sugnaigiul** [reply]
Sure, here are some of the files that i tested (i hosted them on discord CDN because github does not support mp3 uploads 🤦‍♂️) :

File that works straight from the mp3 download : 
https://cdn.discordapp.com/attachments/847423308369887245/1300149807498788884/piri.mp3?ex=671fca5d&is=671e78dd&hm=b9d10f9a4f15b3c4a8a079ffd7bbfebf767ecd3a6b56b394d59e25dcb7236475&

File that did not work at first : 
https://cdn.discordapp.com/attachments/847423308369887245/1300150459570327562/invisibleV1.mp3?ex=671fcaf9&is=671e7979&hm=afb218b328357218fd7a548a16b32ba831a7f6e07b54275e44431590346be9d9&

File that worked after i used a website to remove its metadata : 
https://cdn.discordapp.com/attachments/847423308369887245/1300150991693283329/invisibleV2.mp3?ex=671fcb78&is=671e79f8&hm=7670383f55d9155f5bf27493a61e8d1ff9af31fa066c6b8e2434e5d03a15e4d0&

**Reply by pschatzmann** [reply]
This seems to be a little crazy: usually the meta data is just rund 100 to 200 bytes:  In your case it seems to be a couple of kbytes!
As a work around you can try to set the copy size in the copier to something which is bigger then the metadata....

**Reply by sugnaigiul** [reply]
Ohhh i was not familiar with the normal size of the meta data, that makes more sense.
I will look into all of that.
Thanks a lot for your advice and all these tools !

**Comment by pschatzmann** [comment]
I committed the class MetaDataFilter: You can use this to remove the disturbing metadata.
Just build a chain like this:
```
#include "AudioTools.h"
#include "AudioTools/AudioCodecs/CodecMP3Helix.h"

I2SStream i2s;                           // Output stream
MP3DecoderHelix mp3;                     // Decoder
EncodedAudioStream decoder(i2s, mp3);    // Decoding stream
MetaDataFilter filter(decoder);          // remove metadata

```
and copy the data to the filter.
I tested with your file and it works perfectly...

**Reply by sugnaigiul** [reply]
This is great, thanks a lot !

**Reply by hash6iron** [reply]
Hello @pschatzmann 

Please, Could you tell me how can I do "and copy the data to the filter" in the example above? I have coded a MP3 player, and some mp3 files makes problems with metadata. I'd like to filter it, but I don't achieve to coded the last part of this code.

`
#include "AudioTools.h"
#include "AudioTools/AudioCodecs/CodecMP3Helix.h"

I2SStream i2s;                           // Output stream
MP3DecoderHelix mp3;                     // Decoder
EncodedAudioStream decoder(i2s, mp3);    // Decoding stream
MetaDataFilter filter(decoder);          // remove metadata
`

My mp3 player, is one part of a project dedicated to digital cassettes for 8-bits computers. You can see it here: https://github.com/hash6iron/powadcr

The MP3 player function.

`
void playMP3()
{
    ESP32kit.setSampleRate(AUDIO_HAL_44K_SAMPLES);
    ESP32kit.setSpeakerActive(ACTIVE_AMP);

    SAMPLING_RATE = 44100;
    
    int status = 0;
    int fileSize = 0;
    int fileread = 0;
    int startTime = millis(); 
    int stateWAVplayer = 0;
    int totalFilesIdx = 0;
    int currentFileIdx = 0;

    Boost boost(0.5);

    String txtR = "";


    AudioInfo info(44100, 2, 16);
    // ADPCMDecoder adpcm_decoder(AV_CODEC_ID_ADPCM_IMA_WAV); 
    // WAVDecoder wavdecoder(adpcm_decoder, AudioFormat::ADPCM);
    // WAVDecoder decoder;
    SdSpiConfig sdcfg(PIN_AUDIO_KIT_SD_CARD_CS,SHARED_SPI,SD_SCK_MHZ(SD_SPEED_MHZ),&SPI);
    AudioSourceSDFAT source(FILE_LAST_DIR.c_str(),"mp3",sdcfg);
    AudioKitStream kit;

    // A2DPStream outbt;
    MP3DecoderHelix decoder;
    
    AudioPlayer player(source,kit,decoder);

    Equilizer3Bands eq(kit);
    ConfigEquilizer3Bands cfg_eq;
    
// MetaDataFilter filter(decoder);
     
    auto cfg = kit.defaultConfig(TX_MODE);  
    cfg.bits_per_sample = 16;
    cfg.channels = 2;
    cfg.sample_rate = 44100;
    cfg.buffer_count = 2;
    cfg.buffer_size = 256;

    cfg_eq = eq.defaultConfig();
    cfg_eq.setAudioInfo(cfg); // use channels, bits_per_sample and sample_rate from kit
    cfg_eq.gain_low = EQ_LOW; 
    cfg_eq.gain_medium = EQ_MID;
    cfg_eq.gain_high = EQ_HIGH;
    
    decoder.setOutput(eq);
    kit.begin(cfg);
    kit.setSpeakerActive(ACTIVE_AMP);
 
    
    // setup player
    player.setVolume(1);
    kit.setVolume(MAIN_VOL/100);

    // Iniciamos ecualizador
    eq.begin(cfg_eq);

    // playerbt.setVolume(MAIN_VOL/100);
    player.setAutoNext(false); 

    player.setBufferSize(4096);
    

    if(player.begin())
    {
        LAST_MESSAGE =  "Audio ready to play";

        player.setPath(PATH_FILE_TO_LOAD.c_str());
        fileSize = getWAVfileSize(PATH_FILE_TO_LOAD);

        player.setActive(true);

        // Indicamos cual es el fichero seleccionado
        source.selectStream(FILE_PTR_POS + FILE_IDX_SELECTED - 1);

        // Esto lo hacemos para parar cuando termine el audio.
        int currentIdx = (FILE_PTR_POS + FILE_IDX_SELECTED) - 1;

        // Indicadores
        totalFilesIdx = source.size() + 1;
        currentFileIdx = source.index() + 1;
        
        logln("Path:   " + PATH_FILE_TO_LOAD);
        logln("Source: " + String(source.toStr()));

        FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
        logln("File: " + FILE_LOAD);

        // Esto lo hacemos para coger el bitrate la primera vez
        player.setVolume(0);
        kit.setVolume(0);
        player.copy();
        updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

        // Comenzamos el bucle de control
        while(!EJECT && !REC)
        {          
          
          player.setVolume(1);
          kit.setVolume(MAIN_VOL/100);

          if (EQ_CHANGE)
          {
              EQ_CHANGE = false;
  
              cfg_eq.gain_low = EQ_LOW; 
              cfg_eq.gain_medium = EQ_MID;
              cfg_eq.gain_high = EQ_HIGH;
  
              eq.begin(cfg_eq);
          }
  
          switch (stateWAVplayer)
          {
            case 0:
                // Fichero a la espera
                if (PLAY)
                {
                    LAST_MESSAGE =  "Playing Audio.";

                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;

                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);   
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    currentIdx = source.index();
                    logln("Current IDX: " + String(currentIdx));

                    tapeAnimationON();
                    stateWAVplayer = 1;
                }
                
                // Control de avance de ficheros                
                if (FFWIND)
                {
                    // Lo hacemos asi porque si avanzamos con player.next() 
                    // da error si se pasa del total
                    int idx = source.index() + 1;
                    // Comprobamos
                    if (idx > source.size())
                    {
                      source.selectStream(0);
                    }
                    else
                    {
                      player.next();
                    }

                    // hmi.writeString("name.txt=\"" + String(source.toStr()) + "\"");
                    fileSize = getWAVfileSize(source.toStr());

                    // Ahora el current index ha cambiado
                    currentIdx = source.index();
  
                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;
                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    FFWIND = false;
                    RWIND = false;
                    fileread = 0;
                }
                else if (RWIND)
                {
                    // Lo hacemos asi porque si avanzamos con player.next() 
                    // da error si se pasa del total
                    int idx = source.index() - 1;
                    // Comprobamos
                    if (idx < 0)
                    {
                      source.selectStream(source.size());
                    }
                    else
                    {
                      player.previous();
                    }

                    // hmi.writeString("name.txt=\"" + String(source.toStr()) + "\"");
                    fileSize = getWAVfileSize(source.toStr());

                    // Ahora el current index ha cambiado
                    currentIdx = source.index();

                    // Informamos
                    // updateIndicators(source.size() + 1, source.index() + 1, fileSize, source.toStr()); 

                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;
                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    FFWIND = false;
                    RWIND = false;
                    fileread = 0;
                }
                
                break;
          
          case 1:
                // Play
                if (PLAY)
                {
                    if (BTN_PLAY_PRESSED)
                    {
                      BTN_PLAY_PRESSED = false;
                      
                      if (disable_auto_wav_stop)
                      {
                          disable_auto_wav_stop = false;
                          LAST_MESSAGE = "Auto-stop playing";
                      }
                      else
                      {
                          disable_auto_wav_stop = true;
                          LAST_MESSAGE = "Continuos playing";
                      }
                    }

                    // Vamos transmitiendo y quedandonos con el numero de bytes transmitidos
                    // eqcpy.copy();
                    fileread += player.copy();

                    // Playing time
                    estimatePlayingTime(fileread, fileSize, decoder.audioInfoEx().bitrate);                 
                    // Mostramos la progresion de la reproduccion
                    PROGRESS_BAR_TOTAL_VALUE = (fileread * 100) / fileSize;
                    PROGRESS_BAR_BLOCK_VALUE = PROGRESS_BAR_TOTAL_VALUE;

                    // Cuando cambie automaticamente al proximo fichero, me paro.
                    int idx = source.index();
                    if (currentIdx != idx)
                    {
                        if (!disable_auto_wav_stop)
                        {
                            LAST_MESSAGE = "Stop playing";
                            stateWAVplayer = 4;
                        }
                        else
                        {
                            // Pasamos a la siguiente pista
                            logln("Current IDX: " + String(currentIdx + 1) + " / " + String(source.size()));

                            // Comprobamos si hemos llegado al final de
                            // la lista de reproduccion (play list)
                            if ((currentIdx + 1) > source.size())
                            {
                              source.selectStream(0);
                              player.begin();

                              logln("Change IDX: " + String(source.index()) + " / " + String(source.size()));
                            }

                            // Indicadores
                            totalFilesIdx = source.size() + 1;
                            currentFileIdx = source.index() + 1;
                            FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);       
                            fileSize = getWAVfileSize(source.toStr());
                            updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);
                            // Actualizamos la referencia de cambio de pista 
                            currentIdx = source.index();  
                            fileread = 0;                                         
                        }
                    }                  
                } 
                else if (PAUSE)
                {
                    tapeAnimationOFF();
                    PAUSE = false;
                    stateWAVplayer = 2;
                }
                else if (STOP)
                {
                    stateWAVplayer = 3;
                    fileread = 0;
                    // Lo ponemos asi para que empiece en CONTINUOS PLAYING
                    // ya que en el if se hace un cambio dependiendo del valor anterior
                    disable_auto_wav_stop = false;
                }
                
                if (FFWIND)
                {
                    // Lo hacemos asi porque si avanzamos con player.next() 
                    // da error si se pasa del total
                    int idx = source.index() + 1;
                    // Comprobamos
                    if (idx > source.size())
                    {
                      source.selectStream(0);
                    }
                    else
                    {
                      player.next();
                    }

                    // hmi.writeString("name.txt=\"" + String(source.toStr()) + "\"");
                    fileSize = getWAVfileSize(source.toStr());

                    // Ahora el current index ha cambiado
                    currentIdx = source.index();

                    // Informamos
                    // updateIndicators(source.size() + 1, source.index() + 1, fileSize, source.toStr()); 

                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;
                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    FFWIND = false;
                    RWIND = false;
                    fileread = 0;
                }
                else if (RWIND)
                {
                    // Lo hacemos asi porque si avanzamos con player.next() 
                    // da error si se pasa del total
                    int idx = source.index() - 1;
                    // Comprobamos
                    if (idx < 0)
                    {
                      source.selectStream(source.size());
                    }
                    else
                    {
                      player.previous();
                    }

                    // hmi.writeString("name.txt=\"" + String(source.toStr()) + "\"");
                    fileSize = getWAVfileSize(source.toStr());

                    // Ahora el current index ha cambiado
                    currentIdx = source.index();

                    // Informamos
                    // updateIndicators(source.size() + 1, source.index() + 1, fileSize, source.toStr()); 

                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;
                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    FFWIND = false;
                    RWIND = false;
                    fileread = 0;
                }
                break;

          case 2:
                // Pausa reanudamos el PLAY
                if (PAUSE || PLAY)
                {
                    tapeAnimationON();
                    stateWAVplayer = 1;
                    PLAY = true;
                    PAUSE = false;
                    // Lo ponemos asi para que empiece en CONTINUOS PLAYING
                    // ya que en el if se hace un cambio dependiendo del valor anterior
                    disable_auto_wav_stop = false;                    
                }             
                break;
          case 3:
                // STOP tape

                if (STOP)
                {
                    LAST_MESSAGE = "Stop playing";
                    tapeAnimationOFF();
                    currentIdx = source.index();  
                    stateWAVplayer = 0;  
                    player.begin();
                    source.selectStream(currentIdx);
                    // updateIndicators(source.size() + 1, source.index() + 1, fileSize, source.toStr()); 

                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;
                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    // Lo ponemos asi para que empiece en CONTINUOS PLAYING
                    // ya que en el if se hace un cambio dependiendo del valor anterior
                    disable_auto_wav_stop = false;                    

                }
                break;
          case 4:
                // Auto-stop

                // Ponemos esta condicion para que no de error 
                // la declaracion "int idx = source.index()"
                if (true)
                {
                    // Mandamos parar el reproductor
                    hmi.verifyCommand("STOP");
                    // Lo indicamos
                    LAST_MESSAGE = "Stop playing";
                    // Paramos la animacion de la cinta
                    tapeAnimationOFF();
                    // Ponemos la pista al principio
                    player.begin();
                    // Nos quedamos en la misma pista
                    source.selectStream(currentIdx);
                    // Actualizamos indicadores
                    // updateIndicators(source.size() + 1, source.index() + 1, fileSize, source.toStr());  

                    // Indicadores
                    totalFilesIdx = source.size() + 1;
                    currentFileIdx = source.index() + 1;
                    FILE_LOAD = String(source.toStr()).substring(String(source.toStr()).lastIndexOf("/")+1);
                    
                    updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);

                    // hmi.writeString("name.txt=\"" + String(source.toStr()) + "\""); 
                    // Reseteamos el contador de progreso                  
                    fileread = 0;
                    // Saltamos al estado de reposo  
                    stateWAVplayer = 0;  
                    // Lo ponemos asi para que empiece en CONTINUOS PLAYING
                    // ya que en el if se hace un cambio dependiendo del valor anterior
                    disable_auto_wav_stop = false;                    
                }
                break;

          default:
            break;
          }

          // Actualizamos el indicador
          if ((millis() - startTime) > 4000)
          {
            updateIndicators(totalFilesIdx, currentFileIdx, fileSize, decoder.audioInfoEx().bitrate, FILE_LOAD);
            startTime = millis();
          }       
          
        }
        
        tapeAnimationOFF(); 
        player.end();
        moveDirection = 1;
        posRotateName = 0;        
    }
    else
    {
      logln("Error player initialization");
      STOP=true;
    } 

    // Recupero la SD para el resto de procesadores.
    if (!sdf.begin(sdcfg)) 
    {
      logln("Error recovering spi initialization");
    }  
}
`

thanks in advanced

**Reply by pschatzmann** [reply]
This has [already been answered](https://github.com/pschatzmann/arduino-audio-tools/discussions/1842)

**Reply by hash6iron** [reply]
Hi again. I just saw this comment, https://github.com/pschatzmann/arduino-audio-tools/discussions/1842

I try to do it using MetadataFilterDecoder

**Reply by hash6iron** [reply]
Thanks so much. 
You are doing an excellent work with this library.

**Reply by hash6iron** [reply]
Hello @pschatzmann .

I have used MetadataFilterDecoder but I have observed that some MP3 with Variable bit rate not playing properly, in a concretely moments of playing not ear nothing and later yes. If I left to use MetadaFilterDecoder all is ok.

Do you have what could be the problem?

**Reply by pschatzmann** [reply]
I am not aware of any issues: please provide some example mp3 which have a problem

**Reply by hash6iron** [reply]
Yes. Here attached is one of these files. it's VBR mp3. When I apply metadata filtering, near of the begining of file the playing is not properly.
Thanks in advance
[2. Horsell common y el rayo de fuego.mp3.zip](https://github.com/user-attachments/files/18395268/2.Horsell.common.y.el.rayo.de.fuego.mp3.zip)

**Comment by pschatzmann** [comment]
Are you sure that you just give the decoder not enough buffers with this insane bitrate ?
![image](https://github.com/user-attachments/assets/2139ca52-e53f-4528-83ff-00ef52fea68b)

**Reply by hash6iron** [reply]
The mp3 playing very good without metadatafilter. But if I apply metadatafilter not playing good. But I'm going to try with other buffer size, currently is sized at 4096 bytes.

**Reply by hash6iron** [reply]
I have tried with 40KB, 80KB, 256KB, 512KB and the problem continue, always in the same position of playing (for each one of the mp3's)

**Reply by pschatzmann** [reply]
The MetaDataFilterDecoder did not forward the output, so helix did not know where to send the decoded output.
I also added this class to the common functionality, so that no separate include is necessary.

These [corrections have been committed](https://github.com/pschatzmann/arduino-audio-tools/commit/212996ff78c8dbc43d168bfef64d9b15a9ff876c)

**Reply by hash6iron** [reply]
Is necessary do any modification or is transparent to my own code?

**Reply by hash6iron** [reply]
I have updated the library and now works very good. Thanks so much.
